src/main/kotlin/io/bazel/kotlin/compiler/BuildToolsAPICompiler.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2018 The Bazel Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.bazel.kotlin.compiler
+
+import org.jetbrains.kotlin.buildtools.api.CompilationResult
+import org.jetbrains.kotlin.buildtools.api.CompilationService
+import org.jetbrains.kotlin.buildtools.api.ExperimentalBuildToolsApi
+import org.jetbrains.kotlin.buildtools.api.ProjectId
+import org.jetbrains.kotlin.buildtools.api.SourcesChanges
+import org.jetbrains.kotlin.buildtools.api.jvm.ClasspathSnapshotBasedIncrementalCompilationApproachParameters
+import org.jetbrains.kotlin.cli.common.ExitCode
+import org.jetbrains.kotlin.incremental.createDirectory
+import java.io.File
+import java.nio.file.Paths
+import java.util.UUID
+
+fun stripNonKotlinArgument(arg: String, args: MutableList<String>): String? {
+    val index = args.indexOf(arg)
+    if (index < 0 || index + 1 >= args.size) return null
+    return args.removeAt(index + 1).also { args.removeAt(index) }
+}
+
+@Suppress("unused")
+class BuildToolsAPICompiler {
+  @OptIn(ExperimentalBuildToolsApi::class)
+  fun exec(
+    errStream: java.io.PrintStream,
+    vararg args: String,
+  ): ExitCode {
+    System.setProperty("zip.handler.uses.crc.instead.of.timestamp", "true")
+
+    var mutableArgs = args.toMutableList()
+
+    // Set of vararg arguments into BuildToolsAPICompiler not to be passed to the compiler.
+    // Allows keeping the reflection-based invocation consistent with BazelK2KVMCompiler
+    val snapshots = stripNonKotlinArgument("-snapshot", mutableArgs)
+      ?.split(":")
+      ?.map { File(it) }
+      .orEmpty()
+
+    // A manufactured strategy for delineating running KSP plugins and compiling Kotlin.
+    // label: The Bazel label
+    // incrementalId: Used to delineate Kotlin tasks within a label (i.e. KSP vs Kotlin vs Kapt)
+    val label = stripNonKotlinArgument("-label", mutableArgs)!!
+    val incrementalId = stripNonKotlinArgument("-incremental_id", mutableArgs)!!
+
+    // The incremental directory is roughly analogous to a Gradle transform cache directory. Must be unique per compilation target.
+    val incrementalDirectory = Paths.get("").toAbsolutePath().parent.resolve("_kotlin_incremental/$label/$incrementalId").toFile()
+
+    val kotlinService = CompilationService.loadImplementation(this.javaClass.classLoader!!)
```

Comment:
Is it possible/safe to keep the CompilationService implementation around?

Reply:
Do you mean simply not loading it every compile task? I'm sure that's safe, yeah.